File: Algorithms/动态规划/数位DP/区间内满足限制的数/数位之间相关/连续/LC600.cpp
#include <bits/stdc++.h>
using namespace std;
#define endl "\n"
#define int long long
typedef long long LL;
typedef unsigned long long ULL;
typedef pair<int, int> PII;
const int INF = 0x3f3f3f3f, MOD = 1e9 + 7;

// https://leetcode.cn/problems/non-negative-integers-without-consecutive-ones/

class Solution {
public:
    int findIntegers(int n) {
        int nums[35], f[35][3];
        
        
        int len = 0;
        while (n) nums[++len] = n & 1, n >>= 1;
        function<int(int, int, bool, bool)> dfs = [&](int pos, int cnt, bool isLimit, bool isNum) -> int {
            if (!pos) return isNum;
            if (!isLimit && isNum && f[pos][cnt] != -1) return f[pos][cnt];
            int res = 0;
            if (!isNum) res = dfs(pos - 1, cnt, false, false);
            for (int i = 1 - isNum, up = isLimit ? nums[pos] : 1; i <= up; ++i) {
                int t = i? cnt + 1 : 0;
                if (t > 1) continue;
                res += dfs(pos - 1, t, isLimit && i == nums[pos], true);
            }
            if (!isLimit && isNum) f[pos][cnt] = res;
            return res;
        };
        return dfs(len, 0, true, false) + 1;
    }
};

signed main() {
    system("pause");
    return 0;
}

File: 动态规划/数位DP/区间内满足限制的数/数位之间相关/连续/lc600.py
from functools import cache

# https://leetcode.cn/problems/non-negative-integers-without-consecutive-ones/?envType=daily-question&envId=2024-08-05


class Solution:
    def findIntegers(self, n: int) -> int:
        s = bin(n)[2:]

        @cache
        def dfs(i, pre, isLimit, isNum):
            if i == len(s):
                return isNum
            res = 0
            if not isNum:
                res = dfs(i + 1, pre, False, False)
            lo = 1 - isNum
            hi = int(s[i]) if isLimit else 1
            for x in range(lo, hi + 1):
                if pre and x:
                    continue
                res += dfs(i + 1, x, isLimit and x == hi, True)
            return res

        return dfs(0, 0, True, False) + 1
